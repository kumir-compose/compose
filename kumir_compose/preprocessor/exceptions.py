from kumir_compose.preprocessor.tokens import DIRECTIVES


class LexerException(Exception):
    """Base class for lexer exceptions."""

    def __init__(
            self,
            message: str,
            filename: str,
            source: str,
            line: int,
            char: int
    ) -> None:
        """Create exception."""
        self.message = message
        self.filename = filename
        self.src = source
        self.line = line
        self.char = char

    @property
    def formatted_message(self) -> str:
        """Create beautiful representation of exception."""
        line_str = self.src.splitlines()[self.line - 1]
        line_str_prefix = f"{self.line} |  "
        pointer_margin = " " * (len(line_str_prefix) + self.char)
        return (
            f"{self.message}\n"
            f"At file {self.filename}, line {self.line}, char {self.char}:\n"
            f"{line_str_prefix}{line_str}\n"
            f"{pointer_margin}^"
        )

    def __str__(self):
        """Convert to string repr."""
        return self.formatted_message


class UnexpectedCharacterException(LexerException):
    """Raised when encountered unexpected character."""

    def __init__(
            self,
            got: str,
            expected: str,
            filename: str,
            source: str,
            line: int,
            char: int
    ) -> None:
        """Create exception and message."""
        got = f"'{got}'" if got else "END OF FILE"
        super().__init__(
            f"Unexpected character {got}, expected '{expected}'",
            filename,
            source,
            line,
            char,
        )


class UnknownDirectiveException(LexerException):
    """Raised when encountered unexpected directive."""

    def __init__(
            self,
            got: str,
            filename: str,
            source: str,
            line: int,
            char: int
    ) -> None:
        """Create exception and message."""
        super().__init__(
            (
                f"Unexpected character '{got}', "
                f"expected one of {DIRECTIVES.keys()}"
            ),
            filename,
            source,
            line,
            char,
        )
