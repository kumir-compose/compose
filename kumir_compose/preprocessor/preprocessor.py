from collections.abc import Collection, Sequence
from pathlib import Path

from attrs import frozen

from kumir_compose.preprocessor.exceptions import (
    IncludeFileNotFoundException,
    UnexpectedTokenException,
)
from kumir_compose.preprocessor.lexer import scan_file
from kumir_compose.preprocessor.tokens import Token, TokenType


@frozen
class Condition:
    name: str
    expected: bool


class Preprocessor:
    def __init__(
            self,
            filename: str,
            source: str,
            tokens: Sequence[Token],
            *,
            lookup_paths: Collection[str],
            encoding: str | None = None
    ) -> None:
        self._tokens = list(tokens)
        self._pos = 0
        self._filename = filename
        self._source = source
        self._macro_table: dict[str, Sequence[Token]] = {}
        self._lookup_paths = set(lookup_paths)
        self._lookup_paths.add("")
        self._encoding = encoding
        self._condition_stack = []
        self._included = set()

    @property
    def _at_end(self) -> bool:
        return self._pos >= len(self._tokens)

    def _next(self) -> Token | None:
        if self._at_end:
            return None
        self._pos += 1
        return self._tokens[self._pos - 1]

    @property
    def _peek(self) -> Token | None:
        if self._at_end:
            return None
        return self._tokens[self._pos]

    @property
    def _previous(self) -> Token:
        return self._tokens[self._pos - 1]

    def _match(
            self,
            expected_type: TokenType,
    ) -> Token | None:
        if self._at_end:
            return None
        if self._peek.type == expected_type:
            return self._next()
        return None

    def _error(self, exc_type, offset: int = 0, *args, **kwargs):
        kwargs["source"] = self._source
        kwargs["filename"] = self._filename
        kwargs["token"] = self._tokens[
            min(len(self._tokens) - 1, max(self._pos + offset, 0))
        ]
        raise exc_type(*args, **kwargs)

    def _require(
            self,
            expected_type: TokenType,
            *,
            message: str | None = None,
    ) -> Token:
        token = self._match(expected_type)
        if token is None:
            self._error(
                UnexpectedTokenException,
                expected=expected_type,
                got=self._peek,
                message=message,
            )
        return token

    def _consume_spaces(self) -> None:
        while self._match(TokenType.SPACE): ...

    def process(self):
        """Run preprocessor."""
        processed = ""
        while not self._at_end:
            token = self._next()
            if token.type == TokenType.ENDIF:
                self._end_conditional()
                continue
            if not self._do_conditions_match():
                continue
            match token.type:
                case TokenType.DEFINE:
                    self._define_macro()
                case TokenType.UNDEF:
                    self._undef_macro()
                case TokenType.INCLUDE:
                    self._include()
                case TokenType.IFDEF:
                    self._start_conditional(expected=True)
                case TokenType.IFNDEF:
                    self._start_conditional(expected=False)
                case TokenType.ID:
                    new_token = self._process_id(token)
                    if new_token.lexeme != token.lexeme:
                        self._pos -= 1
                    else:
                        processed += new_token.lexeme
                case _:
                    processed += token.lexeme
        return processed

    def _define_macro(self) -> None:
        self._consume_spaces()
        macro_name = self._require(TokenType.ID)
        self._consume_spaces()
        macro_value = self._next()
        if macro_value.type == TokenType.NEWLINE:
            self._macro_table[macro_name.value] = []
            return
        macro_values = [macro_value]
        while self._peek and self._peek.type != TokenType.NEWLINE:
            macro_values.append(self._next())
        self._macro_table[macro_name.value] = macro_values

    def _undef_macro(self) -> None:
        self._consume_spaces()
        macro_name = self._require(TokenType.ID)
        if macro_name.value in self._macro_table:
            self._macro_table.pop(macro_name.value)

    def _include(self) -> None:
        self._consume_spaces()
        path_tok = self._match(TokenType.ID)
        if not path_tok:
            path_tok = self._require(TokenType.STRING)
        found_path = self._find_file(path_tok.value)
        if not found_path:
            self._error(
                IncludeFileNotFoundException,
                incl_file=path_tok.value,
                offset=-2
            )
        abs_path = str(found_path.absolute())
        if abs_path in self._included:
            return
        tokens = scan_file(abs_path, self._encoding)
        self._insert_tokens_here(list(tokens))
        self._included.add(abs_path)

    def _insert_tokens_here(self, tokens: list[Token]) -> None:
        self._tokens = (
            self._tokens[:self._pos] +
            list(tokens) +
            self._tokens[self._pos:]
        )

    def _find_file(self, filename) -> Path | None:
        for root in self._lookup_paths:
            if Path(root, filename).exists():
                return Path(root, filename)
            if Path(root, f"{filename}.kum").exists():
                return Path(root, f"{filename}.kum")
        return None

    def _start_conditional(self, *, expected: bool) -> None:
        self._consume_spaces()
        def_name = self._require(TokenType.ID)
        self._condition_stack.append(Condition(
            name=def_name.value,
            expected=expected
        ))

    def _end_conditional(self) -> None:
        self._condition_stack.pop()

    def _do_conditions_match(self) -> bool:
        for condition in self._condition_stack:
            is_present = condition.name in self._macro_table
            if is_present != condition.expected:
                return False
        return True

    def _process_id(self, token: Token) -> Token:
        if token.lexeme in self._macro_table:
            first_token, *other_tokens = self._macro_table[token.lexeme]
            self._tokens[self._pos - 1] = first_token
            self._insert_tokens_here(other_tokens)
            return first_token
        return token
