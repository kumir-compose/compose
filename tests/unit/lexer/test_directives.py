from kumir_compose.preprocessor.lexer import Lexer
from tests.utils import read_fix_file


def test_directives_parsed(snapshot):
    """Test that directives are parsed."""
    lexer = Lexer("", read_fix_file("directives.kum"))
    tokens = lexer.scan()
    str_repr = "\n".join(map(str, tokens))
    assert str_repr == snapshot
