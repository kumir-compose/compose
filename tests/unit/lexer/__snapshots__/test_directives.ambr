# serializer version: 1
# name: test_directives_parsed
  '''
  TokenType.INCLUDE:|| включить:1:11
  TokenType.SPACE: :1:12
  TokenType.ID:krpc:1:16
  TokenType.NEWLINE:
  :1:17
  TokenType.DEFINE:|| задать:2:9
  TokenType.SPACE: :2:10
  TokenType.ID:ОТЛАДКА:2:17
  TokenType.NEWLINE:
  :2:18
  TokenType.DEFINE:|| задать:3:9
  TokenType.SPACE: :3:10
  TokenType.ID:ПРОДАКШН:3:18
  TokenType.NEWLINE:
  :3:19
  TokenType.DEFINE:|| задать:4:9
  TokenType.SPACE: :4:10
  TokenType.ID:НАПЕЧАТАТЬ:4:20
  TokenType.PUNCTUATION:(:4:21
  TokenType.DEF_ARG:$a$:4:24
  TokenType.PUNCTUATION:,:4:25
  TokenType.SPACE: :4:26
  TokenType.DEF_ARG:$b$:4:29
  TokenType.PUNCTUATION:,:4:30
  TokenType.SPACE: :4:31
  TokenType.DEF_ARG:$c$:4:34
  TokenType.PUNCTUATION:):4:35
  TokenType.SPACE: :4:36
  TokenType.KEYWORD:вывод:4:41
  TokenType.SPACE: :4:42
  TokenType.DEF_ARG:$a$:4:45
  TokenType.PUNCTUATION:,:4:46
  TokenType.SPACE: :4:47
  TokenType.DEF_ARG:$b$:4:50
  TokenType.PUNCTUATION:,:4:51
  TokenType.SPACE: :4:52
  TokenType.DEF_ARG:$c$:4:55
  TokenType.PUNCTUATION:,:4:56
  TokenType.SPACE: :4:57
  TokenType.ID:нс:4:59
  TokenType.NEWLINE:
  :4:60
  TokenType.IFDEF:|| еслизад:5:10
  TokenType.SPACE: :5:11
  TokenType.ID:ОТЛАДКА:5:18
  TokenType.NEWLINE:
  :5:19
  TokenType.KEYWORD:вывод:6:5
  TokenType.SPACE: :6:6
  TokenType.STRING:"Привет, КуМир-Композитор!":6:33
  TokenType.PUNCTUATION:,:6:34
  TokenType.SPACE: :6:35
  TokenType.ID:нс:6:37
  TokenType.NEWLINE:
  :6:38
  TokenType.ID:НАПЕЧАТАТЬ:7:10
  TokenType.PUNCTUATION:(:7:11
  TokenType.STRING:"а":7:14
  TokenType.PUNCTUATION:,:7:15
  TokenType.SPACE: :7:16
  TokenType.STRING:"б":7:19
  TokenType.PUNCTUATION:,:7:20
  TokenType.SPACE: :7:21
  TokenType.STRING:"в":7:24
  TokenType.PUNCTUATION:):7:25
  TokenType.NEWLINE:
  :7:26
  TokenType.ENDIF:|| конецесли:8:12
  TokenType.NEWLINE:
  :8:13
  TokenType.IFNDEF:|| еслинезад:9:12
  TokenType.SPACE: :9:13
  TokenType.ID:ПРОДАКШН:9:21
  TokenType.NEWLINE:
  :9:22
  TokenType.KEYWORD:лит:10:3
  TokenType.SPACE: :10:4
  TokenType.ID:СЕКРЕТНЫЙ КЛЮЧ:10:19
  TokenType.SPACE: :10:20
  TokenType.PUNCTUATION:=:10:21
  TokenType.SPACE: :10:22
  TokenType.STRING:"12345678":10:32
  TokenType.NEWLINE:
  :10:33
  TokenType.ENDIF:|| конецесли:11:12
  TokenType.NEWLINE:
  :11:13
  TokenType.UNDEF:|| забыть:12:9
  TokenType.SPACE: :12:10
  TokenType.ID:ОТЛАДКА:12:17
  TokenType.NEWLINE:
  :12:18
  TokenType.UNDEF:|| забыть:13:9
  TokenType.SPACE: :13:10
  TokenType.ID:ПРОДАКШН:13:18
  TokenType.NEWLINE:
  :13:19
  '''
# ---
